staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -38,10 +39,109 @@ import (
 	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
 	"k8s.io/apiserver/pkg/endpoints/metrics"
 	endpointsrequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storage"
 
-	klog "k8s.io/klog/v2"
+	"k8s.io/klog/v2"
 )
 
+// watchListEncoder performs encoding of
```

Comment:
I've added a bunch of comments below, but thinking through them more, I would like to revisit the approach a bit more.

We already have something that is called `watchEncoder`.
`watchEncoder` explicitly get `watch.Event` as a parameter, which would allow us to be more precise about operations we're doing and not risking some failure modes that I asked about below.

So I think that instead of this approach, I would like to:
- extext `watchEncoder` directly with the logic that you're adding
- the logic you're adding should then be gated only for events of type `Bookmark`
- for those events, we then explicitly don't cache the response (even if the object would be a CacheableObject`

I believe this would allow us for much simpler and cleaner approach.

Reply:
ok, I think I had a PR for `watchEncoder` (https://github.com/kubernetes/kubernetes/pull/127411).

Reply:
np, I understand that sometimes finding a good solution takes time.
I'm going to open a new PR.

Reply:
I've just opened https://github.com/kubernetes/kubernetes/pull/127587, PTAL.